src/functions.jl

```diff
@@ -5,33 +5,22 @@
 
 Inverse of `sqrt(x)` for non-negative `x`.
 """
-square(x) = x^2
-function square(x::Real)
-    x < zero(x) && throw(DomainError(x, "`square` is defined as the inverse of `sqrt` and can only be evaluated for non-negative values"))
+function square(x)
+    if is_real_type(typeof(x)) && x < zero(x)
+        throw(DomainError(x, "`square` is defined as the inverse of `sqrt` and can only be evaluated for non-negative values"))
+    end
     return x^2
 end
 
 
-function invpow2(x::Real, p::Integer)
```

Comment:
Why don't we keep the dispatch on Integer? Seems nicer than doing `if isinteger`.

Reply:
You're right - also, using `isinteger` will also allow floats that are exact integers. We just have to make sure it's type stable.